store/store.go

```diff
@@ -437,24 +486,26 @@ func (s *Store) doRequest(client *http.Client, reqOptions *requestOptions, user
 
 	wwwAuth := resp.Header.Get("WWW-Authenticate")
 	if resp.StatusCode == 401 && user != nil && strings.Contains(wwwAuth, "needs_refresh=1") {
-		// close previous response, refresh and retry
+		// close previous response, refresh user and retry
 		resp.Body.Close()
-		err = refreshMacaroon(user)
+		err = s.refreshUser(user)
 		if err != nil {
 			return nil, err
 		}
-		if s.authContext != nil {
-			err = s.authContext.UpdateUser(user)
-			if err != nil {
-				return nil, err
-			}
-		}
-		req, err := s.newRequest(reqOptions, user)
+		// retry
+		return s.doRequest(client, reqOptions, user)
+
+	} else if resp.StatusCode == 401 && strings.Contains(wwwAuth, "refresh_device_session=1") {
```

Comment:
You mean do the whole refresh dance for both (device and user macaroons) when we get an expired response?
The expiration time can be very different in each case, so we may end refreshing a macaroon much more often than needed (for ex., user discharge was originally discussed to be refreshed every 2hs). In any case, it should be simple to rework this if we prefer to do a full macaroons refresh.